server/session/handler_book_edit.go

```diff
@@ -0,0 +1,74 @@
+package session
+
+import (
+	"fmt"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/sandertv/gophertunnel/minecraft/protocol/packet"
+	"golang.org/x/exp/slices"
+)
+
+// BookEditHandler handles the BookEdit Packet
+type BookEditHandler struct{}
+
+// Handle ...
+func (b BookEditHandler) Handle(p packet.Packet, s *Session) error {
+	pk := p.(*packet.BookEdit)
+
+	it, err := s.inv.Item(int(pk.InventorySlot))
+	if err != nil {
+		return fmt.Errorf("invalid inventory slot index %v", pk.InventorySlot)
+	}
+	book, ok := it.Item().(item.BookAndQuill)
+	if !ok {
+		return fmt.Errorf("inventory slot %v does not contain a writable book", pk.InventorySlot)
+	}
+
+	page := int(pk.PageNumber)
+	if page >= 50 || page < 0 {
+		return fmt.Errorf("page number %v is out of bounds", pk.PageNumber)
+	}
+	if len(pk.Text) > 256 {
+		return fmt.Errorf("text can not be longer than 256 bytes")
+	}
+
+	slot := int(pk.InventorySlot)
+	switch pk.ActionType {
+	case packet.BookActionReplacePage:
+		book = book.SetPage(page, pk.Text)
+	case packet.BookActionAddPage:
+		if len(book.Pages) >= 50 {
+			return fmt.Errorf("unable to add page beyond 50")
+		}
+		if page >= len(book.Pages) && page <= len(book.Pages)+2 {
```

Comment:
I have not gotten to testing yet but I assume this is meant for inserting a page at the very end in which case this should probably be simplified to:

```suggestion
		if page == len(book.Pages) {
```

Reply:
Completely disregard this